py/src/braintrust/logger.py

```diff
@@ -1121,6 +1147,65 @@ def __exit__(self, type, value, callback):
         self.end()
 
 
+class _ObjectLoader:
+    def __init__(self, object_type: str, object_id: id, version: "str | int" = None):
+        self.finished = False
+        self.object_type = object_type
+        self.object_id = object_id
+
+        self._fetched_data = None
+
+        self._pinned_version = None
+        if version is not None:
+            try:
+                self._pinned_version = int(version)
+                assert self._pinned_version >= 0
+            except (ValueError, AssertionError):
+                raise ValueError(f"version ({version}) must be a positive integer")
+
+    def fetch(self):
+        self._check_not_finished()
+
+        for record in self.fetched_data:
+            yield {
+                "id": record.get("id"),
+                "input": json.loads(record.get("input") or "null"),
+                "output": json.loads(record.get("output") or "null"),
+                "metadata": json.loads(record.get("metadata") or "null"),
+            }
+
+        self._clear_cache()
+
+    @property
+    def fetched_data(self):
+        self._check_not_finished()
+        if not self._fetched_data:
+            resp = log_conn().get(
+                f"object/{self.object_type}",
+                params={"id": self.object_id, "fmt": "json", "version": self._pinned_version},
+            )
+            response_raise_for_status(resp)
+
+            self._fetched_data = [json.loads(line) for line in resp.content.split(b"\n") if line.strip()]
+        return self._fetched_data
+
+    def _clear_cache(self):
+        self._check_not_finished()
+        self._fetched_data = None
+
+    @property
+    def version(self):
+        self._check_not_finished()
+        if self._pinned_version is not None:
+            return self._pinned_version
+        else:
+            return max([int(record.get(TRANSACTION_ID_FIELD, 0)) for record in self.fetched_data] or [0])
+
+    def _check_not_finished(self):
+        if self.finished:
+            raise RuntimeError(f"Cannot invoke method on finished {self.object_type}")
+
+
```

Comment:
Are we missing a method to actually set `finished = True`. Or otherwise is there any need to have a `finished` thing for an object which just reads data?